### PR TITLE
Add scatter_nd cpu

### PR DESCRIPTION
**Pull request by BowenBao**
Related ONNX PR https://github.com/onnx/onnx/pull/2220 (Update: ONNX PR is merged).

~~This PR depends on https://github.com/microsoft/onnxruntime/pull/1756 which bumps the ONNX version.~~ Merged

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      auto* dst = output_tensor->template MutableData<std::string>();
+      std::copy(str_begin, str_end, dst);
+    } else {
+      memcpy(dst_base, src_base, input_tensor->SizeInBytes());
+    }
+  }
+
+  std::vector<int64_t> element_counts(last_indice_dimension, 0LL); // Number of elements for each input dimension
+
+#ifdef USE_OPENMP
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/tensor/scatter_nd.cc]: Do we still use it?

File: onnxruntime/test/python/onnx_backend_test_series.py [hunk 2]
```diff
@@ -124,6 +124,14 @@ def create_backend_test(testname=None):
                                  '^test_range_float_type_positive_delta_expanded_cpu.*',
                                  '^test_range_int32_type_negative_delta_cpu.*',
                                  '^test_range_int32_type_negative_delta_expanded_cpu.*',
+                                 '^test_clip_cpu.*',
```
Comment (tianleiwu) [inline onnxruntime/test/python/onnx_backend_test_series.py]: why change clip tests in scatter_nd change?

Comment (BowenBao) [reply onnxruntime/test/python/onnx_backend_test_series.py]: this seems to be outdated. I guess it's a result of multiple rebase that has happened. I couldn't find this line of change in the "Files changed" tab.

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 1 above)
Comment (BowenBao) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc]: I'm not sure... Do you recommend removing it altogether with the below `omp parallel for` as well?

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  ORT_ENFORCE(indice_tensor != nullptr);
+  ORT_ENFORCE(update_tensor != nullptr);
+
+  auto input_shape   = input_tensor->Shape();
+  auto indice_shape  = indice_tensor->Shape();
+  auto update_shape  = update_tensor->Shape();
+  if (indice_shape.NumDimensions() == 0) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+      "indices tensor must has rank larger than 0");
+  }
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/tensor/scatter_nd.cc]: ScatterND takes three **inputs data tensor of rank r >= 1**, indices tensor of rank q >= 1, and **updates tensor of rank q + r - indices.shape[-1] - 1**.  Shall we check input_shape and update_shape as well?

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 1 above)
Comment (BowenBao) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc]: it seems still used in some other places https://github.com/microsoft/onnxruntime/blob/02c122d6e478d69ee97b3ecdd0aac5cd4f76ba8c/onnxruntime/contrib_ops/cpu/gather_nd.cc#L48

https://github.com/microsoft/onnxruntime/blob/e769617f75ef2ecd68cc6186998b0f6ba3280597/onnxruntime/core/providers/cpu/nn/pool.cc#L48

File: onnxruntime/core/providers/cpu/tensor/scatter_nd.cc (hunk 3 above)
Comment (BowenBao) [reply onnxruntime/core/providers/cpu/tensor/scatter_nd.cc]: Thanks for the catch. Added checks for input_shape. It is possible for update rank to be zero.

**Comment by BowenBao** [comment]
cc @yuslepukhin @tianleiwu - please review

**Comment by BowenBao** [comment]
rebased with master to resolve conflict ..

**Pull request merged**